### ISSUE TITLE
Automate validating source link information

### ISSUE THREAD

**Issue by shawnro** [dotnet/arcade#1873]
In the past we have accidentally included links to dnceng instead of github.  As of now, someone has to do a manual check to ensure all of these are correct.  This check should be something that we can automate and track continually

**Comment by markwilkie** [comment]
@tmat - any thoughts on this?

**Comment by danmosemsft** [comment]
My manual step was pretty dumb, along the lines of
```
c:\temp>\t\strings System.Text.Encoding.pdb | \t\grep documents | \t\grep -v devdiv
{"documents":{"/home/dan/git/corefx/*":"https://raw.githubusercontent.com/danmosemsft/corefx/ded9b8da999dbabb24abb697b429de73229a16bd/*"}}
```
Perhaps there's some proper API for reading portable PDB's? (We're only talking about managed PDB's here, right?)

**Comment by akoeplinger** [comment]
You can install the [sourcelink](https://github.com/ctaggart/SourceLink) tool via `dotnet tool install --global sourcelink` which has some nice helpers:

```
alexander:~$ sourcelink --help
Source Code On Demand

Usage:  [options] [command]

Options:
  -h|--help  Show help information

Commands:
  print-documents  print the documents stored in the pdb or dll
  print-json       print the Source Link JSON stored in the pdb or dll
  print-urls       print the URLs for each document based on the Source Link JSON
  test             test each URL and verify that the checksums match

Use " [command] --help" for more information about a command.
```

**Comment by danmosemsft** [comment]
@akoeplinger that is much better, thank you. Clearly the way to go.

**Comment by tmat** [comment]
C++ PDBs should also use Source Link.

**Comment by danmosemsft** [comment]
Note that @Anipik is about to merge his change to fix sourcelink for the facades coming out of CoreFX (which was the original problem that spurred this conversation). So if we wanted to add a validation step, it need not wait on that.

**Comment by JohnTortugo** [comment]
AFAIU right now [sourcelink cli tool](https://github.com/ctaggart/SourceLink) can only handle Portable PDBs and the PDBs that we have during publish/release are Windows PDBs. I'm still investigating.

**Comment by tmat** [comment]
You could update the cli tool to support Windows PDBs. 

I don't think we should be building Windows PDBs directly though. We should always build Portable PDBs and convert them to Windows PDBs. The converter preserves SourceLink, so validating the Portable PDBs is sufficient.

**Comment by JohnTortugo** [comment]
Thanks @tmat . One question, the symbol uploader supports converting portable pdbs to windows pdbs, can't we disable the pdb conversion during build and only do the conversion during publishing?

**Comment by tmat** [comment]
We can't since it doesn't support embedded PDBs yet. Even if it did, I don't think that's the right place to convert actually. We should be converting during build.

**Comment by JohnTortugo** [comment]
> The converter preserves SourceLink, so validating the Portable PDBs is sufficient.

@tmat so do you think doing this validation during the build is the way to go? I.e., do the validation during the build (before conversion to win pdb) instead of during publishing ?

**Comment by tmat** [comment]
No, we should do the validation on Portable PDBs during publishing. We should also check during publishing that all managed binaries that are being published have Portable or Embedded PDBs.

**Comment by danmosemsft** [comment]
> C++ PDBs should also use Source Link

@anipik were you going to check whether they do - for the native binaries from core repos at least? Wondering whether more work is needed for them.

**Comment by tmat** [comment]
Yes, we should definitely do that. I guess we won't get around reading native PDBs in SourceLink validation.

**Comment by tmat** [comment]
Now that I'm thinking about it more I don't think we have the necessary APIs in Microsoft.DiaSymReader/Native to read native source link data (they are stored in a different stream within the PDB than managed ones). No problem adding those APIs it's just work. 

So I suggest we separate validation of Source Link Portable PDBs and native (C++ generated) PDBs. Let's make sure our Portable PDBs are good first and later on we can add validation of native PDBs.

**Comment by Anipik** [comment]
> for the native binaries from core repos at least? 

I checked sourcelink for the native dlls that we produce in corefx and it was giving me an error as "Invalid Header"

**Comment by tmat** [comment]
@Anipik How did you check?

**Comment by Anipik** [comment]
so one of the native pdbs that we ship is ```clrcompression.pdb``` so I just ran 
```
C:\git\coreclr>sourcelink test E:\sdkmastersymbols\clrcompression.pdb
Invalid COR20 header signature.
```

**Comment by tmat** [comment]
it's not portable pdb so it's not gonna work

**Comment by JohnTortugo** [comment]
> So I suggest we separate validation of Source Link Portable PDBs and native (C++ generated) PDBs. Let's make sure our Portable PDBs are good first and later on we can add validation of native PDBs.

Created this issue to track the work on Portable PDBs: https://github.com/dotnet/arcade/issues/2512

**Comment by JohnTortugo** [comment]
Moved this to the same release as the blocking issue. @markwilkie @riarenas

**Comment by adiaaida** [comment]
Out of scope for #6418.
